src/librustc_typeck/check/mod.rs

```diff
@@ -1064,6 +1066,29 @@ fn check_fn<'a, 'gcx, 'tcx>(inherited: &'a Inherited<'a, 'gcx, 'tcx>,
     }
     fcx.demand_suptype(span, ret_ty, actual_return_ty);
 
+    // If the termination trait language item is activated, check that the main return type
+    // implements the termination trait.
+    if fcx.tcx.lang_items().termination().is_some() {
```

Comment:
nit: use `if let Some(term_id) = fcx.tcx.lang_items().termination()`, then you won't need to `require` it later